src/wal/src/local_storage_impl/segment.rs

```diff
@@ -454,6 +515,76 @@ impl SegmentManager {
         }
         Ok(())
     }
+
+    pub fn mark_delete_entries_up_to(
+        &self,
+        location: WalLocation,
+        sequence_num: SequenceNumber,
+    ) -> Result<()> {
+        let current_segment_id = self.current_segment.lock().unwrap().lock().unwrap().id;
+        let mut segments_to_remove: Vec<u64> = Vec::new();
```

Comment:
Can we avoid this vec with `HashMap::retain`?

Reply:
I think no if we move `segment.delete()` outside all_segment's lock,